### PR TITLE
CFY-4936 add maintenance mode

### PR DESCRIPTION
**Pull request by nirbir**

### CODE REVIEW

File: cloudify_rest_client/maintenance.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+########
+# Copyright (c) 2014 GigaSpaces Technologies Ltd. All rights reserved
```
Comment (dankilman) [inline cloudify_rest_client/maintenance.py]: 2016

File: cloudify_rest_client/maintenance.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        response = self.api.get(uri)
+        return Maintenance(response)
+
+    def activate(self):
+        """
+        Activates maintenance mode.
+
+        :return:
+        """
+        uri = '/maintenance/{0}'.format('activate')
```
Comment (dankilman) [inline cloudify_rest_client/maintenance.py]: no need for format `/maintenance/activate`

File: cloudify_rest_client/maintenance.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        response = self.api.post(uri)
+        return Maintenance(response)
+
+    def deactivate(self):
+        """
+        Deactivates maintenance mode.
+
+        :return:
+        """
+        uri = '/maintenance/{0}'.format('deactivate')
```
Comment (dankilman) [inline cloudify_rest_client/maintenance.py]: same here

**Comment by dankilman** [comment]
I addition to the comments made inline, there should also be two exceptions added to exceptions.py with a matching `ERROR_CODE`s and they should be added to the `ERROR_MAPPING` dict. Moreover, the unit tests in manager-rest should be updated to use them (i.e. `try/except` these specific types instead of the base `CloudifyClientError`)

File: cloudify_rest_client/maintenance.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+class Maintenance(dict):
+
+    def __init__(self, deployment):
+        self.update(deployment)
+
+    @property
+    def status(self):
+        """
+        :return: maintenance mode's status.
```
Comment (dankilman) [inline cloudify_rest_client/maintenance.py]: update comment to specify what are the expected return values.

File: cloudify_rest_client/maintenance.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+
+class Maintenance(dict):
+
+    def __init__(self, deployment):
```
Comment (dankilman) [inline cloudify_rest_client/maintenance.py]: `__init__(self, deployment)` should not be deployment

File: cloudify_rest_client/exceptions.py [hunk 6]
```diff
@@ -181,6 +181,30 @@ def __init__(self, message, server_traceback=None,
             message, server_traceback, status_code, error_code)
 
 
+class MaintenanceModeActiveError(CloudifyClientError):
+    """
+    Raised when a call has been blocked due to maintenance mode being active.
+    """
+    ERROR_CODE = 'maintenance_mode_active'
+
+    def __init__(self, message, server_traceback=None,
```
Comment (nirbir) [inline cloudify_rest_client/exceptions.py]: remove

**Pull request merged**